### PR TITLE
Request pooled txs only up to max message size

### PR DESCRIPTION
**Pull request by marcindsobczak**
## Changes

- After receiving `NewPooledTransactionHashesMessage` we are requesting received (and not already known) transactions by sending `GetPooledTransactionsMessage`. The problem is that we are not checking sizes and we can request transactions which have altogether more than max message size limit. Up to `eth67` there was nothing to do with that, but since `eth68`, `NewPooledTransactionHashesMessage` has tx sizes, so we can divide requested transactions to a few `GetPooledTransactionsMessages` with every one under max size limit

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [x] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Requires documentation update

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by LukaszRozmej** [comment]
Should we limit the hashes per message to some heuristic value that should be under the limit in `eth67` and lower to avoid this issue?

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/PooledTxsRequestor.cs [hunk 1]
```diff
@@ -7,6 +7,7 @@
 using Nethermind.Core.Caching;
 using Nethermind.Core.Collections;
 using Nethermind.Core.Crypto;
+using Nethermind.Network.P2P.Subprotocols.Eth.V62.Messages;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/PooledTxsRequestor.cs:10]: this class should change namespace it is in `V65` now

**Review by LukaszRozmej — COMMENTED** [review]
We should remove from `_pendingHashes` all the requested hashes that we didn't get responses for and re-request them again?

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/PooledTxsRequestor.cs (hunk 1 above)
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/PooledTxsRequestor.cs:10]: What do you mean "is in `V65` now"?
`using Nethermind.Network.P2P.Subprotocols.Eth.V62.Messages;` is needed to use const `TransactionsMessage.MaxPacketSize`, `TransactionsMessage` is a base message for transactions and is a part of `V62`. Peers are still sending to us pure `V62` `TransactionsMessages`. What more, when we are receiving `PooledTransactionsMessage`, we are handling it in `V62` handler, same as pure `V62` `Transactions`.

**Comment by marcindsobczak** [comment]
> We should remove from `_pendingHashes` all the requested hashes that we didn't get responses for and re-request them again?

It is a good idea. The problem here is that we are not pairing `GetPooledTransactionsMessage` with received `PooledTransactionsMessage`. It will require bigger refactoring - we would need to know which hashes we sent to which peer in message with which msgID and compare after receiving response. So maybe create dictionary with peer as a key and as a value another dictionary with msgID and list of hashes? And after receiving `PooledTransactionsMessage` get msgID-hashes dictionary by peer, get hashes by msgID, and compare hashes and remove from `_pendingHashes` missing ones.
On the other hand it is not a big issue - we are requesting transactions from the peer which sent hashes to us, so it should have this transactions, at least most of them - it can not send them to us if
1) transaction was included in block in the meantime (so not a problem - we don't need it anymore)
2) transaction was evicted from the pool (so not a big problem, it will not be included in close future)
3) we disconnected in the meantime - rerequesting would be usuful here - we would need to add logic for not receiving response too - if we don't receive response in let's say 10s, then remove all hashes from cache
4) response would be too big - it will not happen for `eth68` and newer and on mainnet, goerli, sepolia etc. we are already using `eth68`, on nethermind-only networks we would move to `eth68` after implementing snap server or with Cancun fork, so `eth67` and older would be deprecated within about ~6months.

Another thing is that our `_pendingHashes` cache has capacity of 16k hashes so is clearing quite often anyway

It would be nice to add, but definitely is too complex to add it in this PR - I created issue for that https://github.com/NethermindEth/nethermind/issues/5889

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/PooledTxsRequestor.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/PooledTxsRequestor.cs:10]: `Nethermind.Network.P2P.Subprotocols.Eth.V65`, this class is used by multiple protocol versions, it is not in the right place.

**Comment by LukaszRozmej** [comment]
We have `RequestId` in eth66, so we can correlate there. Shouldn't be that hard?

**Comment by marcindsobczak** [comment]
> Should we limit the hashes per message to some heuristic value that should be under the limit in `eth67` and lower to avoid this issue?

Pre `eth68` we are not able to avoid it for 100% certainty until we will not request every tx separately and it is not the greatest idea - we don't want to spam that much.
We can make a limit of max 256 txs in one message - we used such limit in the past for sending standard txs - today we have more sophisticated limiting by size, but the old one was working pretty well too.
On the other hand
1. `eth67` and lower are used only on nethermind-only networks, which has way less traffic than ethereum mainnet, and only until snap server will be implemented or Cancun fork will happen, so should be dropped within about ~6months.
2. Pre Cancun blob-txs (so large ones) are not in use, so amount of big, potentially problematic transactions is not that big. We didn't have this limiting logic until today and it is not a big issue for the current state of the network

But the change is simple so I will add it to this PR today

File: src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/PooledTxsRequestor.cs (hunk 1 above)
Comment (marcindsobczak) [reply src/Nethermind/Nethermind.Network/P2P/Subprotocols/Eth/V65/PooledTxsRequestor.cs:10]: Ah, I got it. Yeah, it was initially used only by `eth65`, now `eth66` and `eth68` is using it as well. I will move it to general `Eth` namespace

**Comment by marcindsobczak** [comment]
> We have `RequestId` in eth66, so we can correlate there. Shouldn't be that hard? We can ignore pre eth66, they are obsolete.

Yes, by `msgID` from my repsonse I meant `RequestId`. We can create dictionary
`Dictionary<reqID,hashes[]>`
Add to it when sending `GetPooledTransactionsMessage` and after receiving `PooledTransactionsMessage` compare hashes of received txs with hashes sent and missing ones remove from `_pendingHashes`.
More problematic would be handling timeouts - in case of no response we need to remove all hashes sent in `GetPooledTransactionsMessage` from `_pendingHashes`. That would be harder part.

Anyway, IMO current changes in this PR, so requesting txs up to max limit for `eth68` and up to 256 in `eth67` and below (I will push it soon) are way more important than handling requested and not received transactions. And this feature will probably take 3x more time than initial PR, it will change our behavior significantly so we will need to test it deeply, there will be some edge-cases to cover etc.
It's a good idea for a feature added in the future, I don't like the idea of delaying merging this PR for that reason. Plus it is easier to review 2 smaller PRs than one big

**Comment by LukaszRozmej** [comment]
> > We have `RequestId` in eth66, so we can correlate there. Shouldn't be that hard? We can ignore pre eth66, they are obsolete.
> 
> Yes, by `msgID` from my repsonse I meant `RequestId`. We can create dictionary `Dictionary<reqID,hashes[]>` Add to it when sending `GetPooledTransactionsMessage` and after receiving `PooledTransactionsMessage` compare hashes of received txs with hashes sent and missing ones remove from `_pendingHashes`. More problematic would be handling timeouts - in case of no response we need to remove all hashes sent in `GetPooledTransactionsMessage` from `_pendingHashes`. That would be harder part.
> 
> Anyway, IMO current changes in this PR, so requesting txs up to max limit for `eth68` and up to 256 in `eth67` and below (I will push it soon) are way more important than handling requested and not received transactions. And this feature will probably take 3x more time than initial PR, it will change our behavior significantly so we will need to test it deeply, there will be some edge-cases to cover etc. It's a good idea for a feature added in the future, I don't like the idea of delaying merging this PR for that reason. Plus it is easier to review 2 smaller PRs than one big

We have a class for that `MessageDictionary`

**Review by LukaszRozmej — APPROVED** [review]
Would still track `eth66` messages and responses and handle correctly.

**Comment by marcindsobczak** [comment]
> Would still track `eth66` messages and responses and handle correctly.

Yeah it would be good. But let's make it in separate PR. Issue: https://github.com/NethermindEth/nethermind/issues/5889
